vocabulary-draft/continuous-integration-pipeline-events.md

```diff
@@ -0,0 +1,29 @@
+# Continuous Integration Pipeline Events
+
+These events are related to continuous integration pipelines, this pipelines usually include building, testing, packaging and releasing software artifacts. 
+Due the dynamic nature of Pipelines, most of actual work needs to be queued to happen in a distributed way, hence Queued events are added. 
+Adopters can choose to ignore these events if they don't apply to their use cases. 
+
+A pipeline, in the context of Continuous Integration, is the definition of a set of tasks that needs to be performed to build, test, package and release software artifacts. A pipeline can be instanciated multiple times, for example to build different versions of the same artifact. We are refering to this instance as PipelineRun. It will have a unique Id and it will help us to track the build and release progress on a particular software artifact. 
+
+- **PipelineRun Queued**: a PipelineRun has been schedule to run
+- **PipelineRun Started**: a PipelineRun has started and it is running
+- **PipelineRun Finished**: a PipelineRun has finished it execution, the event will contain the finished status, success, error or failure
+- **Build Queued**: a Build task has been queued, this build process usually is in charge of producing a binary from source code
+- **Build Started**: a Build task has started 
+- **Build Finished**: a Build task has finished, the event will contain the finished status, success, error or failure
+- **Tests Queued**: a Test task has been queued, and it is waiting to be started
+- **Tests Started**: a Test task has started
+- **Tests Finished**: a Test task has finished, the event will contain the finished status, success, error or failure
```

Comment:
@m-linner-ericsson I think that will be good to cover, what would you suggest? Here we should look for a generic way to capture those stages inside the tests.

Reply:
For example in [Eiffel](https://github.com/eiffel-community/eiffel/tree/master/eiffel-vocabulary) we have:

* [Test Case Triggered](https://github.com/eiffel-community/eiffel/blob/master/eiffel-vocabulary/EiffelTestCaseTriggeredEvent.md)
* [Test Case Canceled](https://github.com/eiffel-community/eiffel/blob/master/eiffel-vocabulary/EiffelTestCaseCanceledEvent.md)
* [Test Case Started](https://github.com/eiffel-community/eiffel/blob/master/eiffel-vocabulary/EiffelTestCaseStartedEvent.md)
* [Test Case Finished](https://github.com/eiffel-community/eiffel/blob/master/eiffel-vocabulary/EiffelTestCaseFinishedEvent.md)
* [Test Suite Started](https://github.com/eiffel-community/eiffel/blob/master/eiffel-vocabulary/EiffelTestSuiteStartedEvent.md)
* [Test Suite Finished](https://github.com/eiffel-community/eiffel/blob/master/eiffel-vocabulary/EiffelTestSuiteFinishedEvent.md)

Where a Test Suite can consist of multiple suites. Thus in the example, I stated we could have a suite to cover all the 3 suites.

Could this be something to start with?

Reply:
@m-linner-ericsson totally.. Let me add that.. can you help me to define the meaning of Test Case and Test Suite. 
Is it a test case an individual Test? is it the smallest unit of testing like a Unit Test? Or a batch of Unit Tests? How is it different from a Test Suite? does the Test Suite contain Test Cases?

Reply:
Do these definitions work?

**Test Case**: The smallest unit of testing that the user wants to track
**Test Suite**:  a test suite is nothing more or less than a collection of test case executions and/or other test suite executions. The executed test suite may be an ad-hoc transient grouping of test cases that were executed at a particular time or place or for a particular purpose or a persistent entity tracked in a test management system [Taken from [Test Suite Started](https://github.com/eiffel-community/eiffel/blob/master/eiffel-vocabulary/EiffelTestSuiteStartedEvent.md) ]

Reply:
@m-linner-ericsson added to the file. Can you please review?

Reply:
Looks good to me but Andrea had some questions in another thread.

Reply:
> Test Case: The smallest unit of testing that the user wants to track
 
This looks good, thank you. 
A couple of notes:
- tests for which events are generated should not be a single unit test or short running test. We are not trying to create a distributed test runner, and that level of granularity might create very large amounts of events with little value. Different languages use machine parsable files for test results (`junitxml`, `subunit` and friends), and I think that's the right format for the single test level of granularity
- example of tests for which an event should be sent are things like a specific CI job, and a test suite would be the sum of all CI jobs
- test suites should be optional. They can be handy to group tests, but the system may not always be able to tell when a test suite is completed or may not always have a concept of test suite

Reply:
I believe there is a reason to provide events for individual test cases, not just for suites or whole CI jobs. Of course they should be optional to use, and it is most probably not reasonable to send events for each test in a unit test suite, but it could very well be suitable to send an event for each test in a performance test suite for example.
Test suites could probably be optional, enabling test cases to be reported without being connected to a test suite, just as for most other events. This brings another discussion point to the table, that we might need to provide a means to set some 'event policies' or 'event templates' or what we should call them, where we could say that for example "if you want a test reporting tool to be able to monitor your tests your system should send these event types: 'test_suite_started', 'test_suite_finished', 'test_case_triggered', 'test_case_started', 'test_case_finished'"

Reply:
I totally agree with @e-backmark-ericsson on this one. And I believe that the word "optional" is key here. All these events should be picked up only if they are necessary for the one emitting or consuming them. I will be ok with just supporting Test Suites in the first iteration, but I am not against having single unit test events.